src/Adapter/MSTestAdapter.PlatformServices/Services/ExecutionContextService.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+
+namespace Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices;
+
+internal static class ExecutionContextService
+{
+    /// <summary>
+    /// The execution context to use by class. This context will be derived from the assembly level context and shared down
+    /// to the tests of this class. The type used as key is the type of the test class and not the type of the method info.
+    /// </summary>
+    private static readonly ConcurrentDictionary<Type, ExecutionContext?> ClassesExecutionContexts = new();
+
+    /// <summary>
+    /// The execution context to use for the tests (test init, method, test cleanup) of an instance. This context will be derived
+    /// from the test class (if it exists) or the assembly level context. The key is the instance of the test class.
+    /// </summary>
+    private static readonly ConcurrentDictionary<object, ExecutionContext?> InstancesExecutionContexts = new();
+
+    /// <summary>
+    /// As we only support one assembly level context, we store it here.
+    /// </summary>
+    private static ExecutionContext? s_assemblyExecutionContext;
+
+    /// <summary>
+    /// When we execute the action, we need to ensure we are restoring the execution context that was captured in the logical flow of execution.
+    /// After the action is executed we capture the current execution context and save it for the next action to use based on the current execution context scope.
+    ///
+    /// The logical flow of execution is:
+    /// - AssemblyInitialize execution context is saved at "assembly level" and a copy is flown to each ClassInitialize.
+    /// - ClassInitialize execution context is saved at "class level" and a copy is flown to each TestInitialize/TestMethod/TestCleanup.
+    /// - TestInitialize/TestMethod/TestCleanup execution context is mutating the "instance level"
+    /// - ClassCleanup reuses the "class level" execution context.
+    /// - AssemblyCleanup reuses the "assembly level" execution context.
+    /// </summary>
```

Comment:
This comment above would be really nice to have as a mermaid graph on this PR, or as ascii with a single concrete example. 

e.g. I set culture at class level, where it will be visible, and if I mutate it in test, where will the change propagate.

Reply:
Good suggestion, I will do it in the PR and add it in the code (will use a different "graph" for the code comment).